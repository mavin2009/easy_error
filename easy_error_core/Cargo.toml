# easy_error_core/Cargo.toml

[package]
name = "easy_error_core"
version = "0.1.0"
edition = "2021"
description = "Core error type for easy_error"
license = "MIT OR Apache-2.0"
repository = "https://github.com/mavin2009/easy_error"
homepage = "https://github.com/mavin2009/easy_error#readme"
documentation = "https://docs.rs/easy_error_core"

[dependencies]
