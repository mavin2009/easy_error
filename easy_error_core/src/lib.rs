// easy_error_workspace/easy_error_core/src/lib.rs

//! # easy_error_core
//!
//! `easy_error_core` provides the core `EasyError` type used for enhanced error handling.

#[derive(Debug)]
pub struct EasyError {
    source: Box<dyn std::error::Error + Send + Sync>,
    context: String,
}

impl EasyError {
    /// Creates a new `EasyError` with the given source error and context.
    pub fn with_context<E>(source: E, context: &str) -> Self
    where
        E: Into<Box<dyn std::error::Error + Send + Sync>>,
    {
        EasyError {
            source: source.into(),
            context: context.to_string(),
        }
    }

    /// Returns a reference to the source error.
    pub fn source(&self) -> &(dyn std::error::Error + 'static) {
        self.source.as_ref()
    }

    /// Returns the context string.
    pub fn context(&self) -> &str {
        &self.context
    }
}

impl std::fmt::Display for EasyError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}: {}", self.context, self.source)
    }
}

impl std::error::Error for EasyError {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        Some(&*self.source)
    }
}
