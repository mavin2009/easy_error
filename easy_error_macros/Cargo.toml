# easy_error_workspace/easy_error_macros/Cargo.toml

[package]
name = "easy_error_macros"
version = "0.1.0"
edition = "2021"
description = "Procedural macros for the easy_error crate"
license = "MIT OR Apache-2.0"
repository = "https://github.com/mavin2009/easy_error"
homepage = "https://github.com/mavin2009/easy_error#readme"
documentation = "https://docs.rs/easy_error_macros"

[lib]
proc-macro = true

[dependencies]
proc-macro2 = "1.0"
quote = "1.0"
syn = { version = "1.0", features = ["full"] }
easy_error_core = { version = "0.1.0", path = "../easy_error_core" }

[dev-dependencies]
trybuild = "1.0"
easy_error = { version = "0.1.0", path = "../easy_error" }
