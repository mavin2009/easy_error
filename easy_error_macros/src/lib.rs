// easy_error_macros/src/lib.rs

use proc_macro::TokenStream;
use quote::quote;
use syn::{
    parse::{Parse, ParseStream},
    parse_macro_input, Error, Expr, Ident, LitStr, Token,
};

/// Macro to enhance the `?` operator by adding context.
#[proc_macro]
pub fn try_easy(input: TokenStream) -> TokenStream {
    // Parse the input tokens into a syntax tree
    let input = parse_macro_input!(input as TryEasyInput);
    let expr = input.expr;
    let context = input.context;

    // Generate code that adds context to the error using `map_err`.
    let expanded = quote! {
        #expr.map_err(|e| ::easy_error_core::EasyError::with_context(e, #context))?
    };

    TokenStream::from(expanded)
}

/// Struct to parse the input for `try_easy!` macro.
struct TryEasyInput {
    expr: Expr,
    _comma: Token![,],
    context: LitStr,
}

impl Parse for TryEasyInput {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        Ok(TryEasyInput {
            expr: input.parse()?,
            _comma: input.parse()?,
            context: input.parse()?,
        })
    }
}

/// Defines a custom error enum with specified variants.
///
/// # Arguments
///
/// * `name` - The name of the error enum.
/// * `variants` - A list of variant names.
///
/// # Example
///
/// ```rust
/// use my_easy_error_macros::{define_error, try_easy};
/// use my_easy_error_core::EasyError;
///
/// define_error!(MyError, IoError, ParseError);
/// ```
#[proc_macro]
pub fn define_error(input: TokenStream) -> TokenStream {
    // Parse the input tokens into a syntax tree
    let input = parse_macro_input!(input as DefineErrorInput);
    let name = input.name;
    let variants = input.variants;

    // Check if variants are provided
    if variants.is_empty() {
        // Emit a compile-time error if no variants are provided
        return Error::new(name.span(), "define_error! requires at least one variant")
            .to_compile_error()
            .into();
    }

    // Generate enum variants
    let variants_enum = quote! {
        #(#variants),*
    };

    // Generate a single `From` implementation that matches on all variants
    let from_impl = quote! {
        impl From<#name> for EasyError {
            fn from(error: #name) -> Self {
                match error {
                    #(
                        #name::#variants => EasyError::with_context(error, stringify!(#variants)),
                    )*
                }
            }
        }
    };

    // Generate the error type with implementations
    let expanded = quote! {
        #[derive(Debug)]
        pub enum #name {
            #variants_enum
        }

        impl std::fmt::Display for #name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                match self {
                    #(
                        #name::#variants => write!(f, stringify!(#variants)),
                    )*
                }
            }
        }

        impl std::error::Error for #name {}

        #from_impl
    };

    // Convert the expanded code into a TokenStream
    TokenStream::from(expanded)
}

/// Struct to parse the input for `define_error!` macro.
struct DefineErrorInput {
    name: Ident,
    variants: Vec<Ident>,
}

impl Parse for DefineErrorInput {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let name = input.parse()?;
        let mut variants = Vec::new();

        if input.peek(Token![,]) {
            input.parse::<Token![,]>()?;
            while !input.is_empty() {
                let variant = input.parse()?;
                variants.push(variant);
                if input.peek(Token![,]) {
                    input.parse::<Token![,]>()?;
                } else {
                    break;
                }
            }
        }

        Ok(DefineErrorInput { name, variants })
    }
}
