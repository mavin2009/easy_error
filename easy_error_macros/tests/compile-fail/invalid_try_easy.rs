// easy_error_macros/tests/compile-fail/invalid_try_easy.rs

use easy_error_core::EasyError;
use easy_error_macros::try_easy;

fn some_operation() -> Result<(), &'static str> {
    Err("operation failed")
}

fn main() -> Result<(), ()> {
    // Incorrect context type (not a string literal)
    try_easy!(some_operation(), 123);
    Ok(())
}
