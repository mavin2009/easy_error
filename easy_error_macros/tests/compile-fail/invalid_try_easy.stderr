error: expected string literal
  --> tests/compile-fail/invalid_try_easy.rs:12:33
   |
12 |     try_easy!(some_operation(), 123);
   |                                 ^^^

warning: unused import: `easy_error_core::EasyError`
 --> tests/compile-fail/invalid_try_easy.rs:3:5
  |
3 | use easy_error_core::EasyError;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default
